### PR TITLE
working on #35

### PR DESCRIPTION
**Pull request by ershovd**
- extract shaking view into ShakeControlHelper class
- create Shake command and logic for empty view validation
- apply shaking in FlagTaskDialog

### CODE REVIEW

File: new/QualityKitchen.Mobile/Views/ShakeControlHelper.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+		{
+			await view.TranslateTo(view.X + FirstShakeOffset, 0, ShakePhaseDuration);
+			await view.TranslateTo(view.X - FirstShakeOffset, 0, ShakePhaseDuration);
+
+			for (int i = 0; i < 2; i++)
+			{
+				await view.TranslateTo(view.X + SubsequentShakeOffset, 0, ShakePhaseDuration);
+				await view.TranslateTo(view.X - SubsequentShakeOffset, 0, ShakePhaseDuration);
+			}
+		}
```
Comment (j0nscalet) [inline new/QualityKitchen.Mobile/Views/ShakeControlHelper.cs]: Make sure you translate the container back to zero so it's in it's original position.

Add:
```
await View.Translate(0,0,ShakePhaseDuration)
```

File: new/QualityKitchen.Mobile/Pages/FlagTaskDialog.xaml.cs [hunk 2]
```diff
@@ -33,5 +34,13 @@ protected override void OnAppearing ()
 		{
 			reasonTextBox.Focus ();
 		}
+
+		private async Task HandleShake()
+		{
```
Comment (j0nscalet) [inline new/QualityKitchen.Mobile/Pages/FlagTaskDialog.xaml.cs]: See my comment on `IsValidInput`, I would pass a parameter from the ViewModel to the View (Page) and use that to determine which control to shake. It keeps the View decouple with the VM. 😎

File: new/QualityKitchen.Mobile/ViewModels/FlagTaskViewModel.cs [hunk 3]
```diff
@@ -57,13 +58,25 @@ private void HandleCloseCommand ()
...
 			// TODO Use a activity spinner in the button here? If not show/hide overlay			
-			await _taskFlaggedCallback (Task.Id, _employeeId, Reason, Details);
+			await _taskFlaggedCallback(Task.Id, _employeeId, Reason, Details);
 
 			CloseCommand.Execute (null);
 		}
 
+		private bool IsValidInput()
+		{
+			if (string.IsNullOrEmpty(Reason) || string.IsNullOrEmpty(Details))
```
Comment (j0nscalet) [inline new/QualityKitchen.Mobile/ViewModels/FlagTaskViewModel.cs]: As I mentioned above, call a parameter to the View (Page) and use that to determine which control to shake.

File: new/QualityKitchen.Mobile/Views/ShakeControlHelper.cs (hunk 1 above)
Comment (ershovd) [inline new/QualityKitchen.Mobile/Views/ShakeControlHelper.cs]: I'll check that, but from I've understand this is already taken in account by +- calls

File: new/QualityKitchen.Mobile/Pages/FlagTaskDialog.xaml.cs (hunk 2 above)
Comment (ershovd) [inline new/QualityKitchen.Mobile/Pages/FlagTaskDialog.xaml.cs]: yes, this is more clear approach. except that we will be to calling the command twice (which is fine, I guess).

**Comment by ershovd** [comment]
@j0nscalet I've implemented `Effect` and `EffectBehavior` as you suggested. Regarding the passing parameter from viewmodel to actual page (on which control to shake) - this is done not very clear way, using flag enum.

File: new/QualityKitchen.Mobile/Views/Effects/EffectBehavior.cs [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+{
+	public class EffectBehavior : Behavior<View>
+	{
+		public const string GroupName = "QualityKitchen";
+		public const string EffectName = "ShakeEffect";
+
+		public static readonly BindableProperty GroupProperty = BindableProperty.Create("Group", typeof(string), typeof(EffectBehavior), null);
+		public static readonly BindableProperty NameProperty = BindableProperty.Create("Name", typeof(string), typeof(EffectBehavior), null);
+
+		public string Group
```
Comment (j0nscalet) [inline new/QualityKitchen.Mobile/Views/Effects/EffectBehavior.cs]: Can you include a comment with some details on what Group property does? It's not clear to me.

File: new/QualityKitchen.Mobile/Views/Effects/ShakingControls.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+using System;
+
+namespace QualityKitchen.Mobile
+{
+	[Flags]
+	public enum ShakingControls
```
Comment (j0nscalet) [inline new/QualityKitchen.Mobile/Views/Effects/ShakingControls.cs]: Can we refactor the logic to not use an Enum here? It doesn't bode well for the future to update the Enum for any new control we wanted to add. 

Thoughts?

**Comment by j0nscalet** [comment]
@ershovd functionality is good. There are some things that jumped out at me when I glanced at the code. Look out for those in a bit.

File: new/QualityKitchen.Mobile/Views/Effects/ShakingControls.cs (hunk 5 above)
Comment (ershovd) [inline new/QualityKitchen.Mobile/Views/Effects/ShakingControls.cs]: @j0nscalet I don't like idea of the enum as well. One solution I see is to use `FindByName` method in view (and pass this name as parameter in Viewmodel Excecute command.
Another approach can be to have an event with parameter View - but again, we need to determine which view to select in viewmodel.
I guess I will proceed with FindByName approach

**Comment by ershovd** [comment]
@j0nscalet ok, that make sense. I will remove the effect and make it to a view extension, so that it can be called in page. I will also remove the enum logic as I mention in previous comment. 
My estimate is 1h for that or less.

**Comment by ershovd** [comment]
@j0nscalet I've implemeted according to suggestions and with FindByName approach. It actually took me 20m.

**Comment by j0nscalet** [comment]
@ershovd You're a rockstar. I'll take a look, thanks Dmitry.

**Pull request merged**